### PR TITLE
feat(python): Add syntactic sugar to `DataFrame.filter`

### LINKED ISSUE #11642: Add .locate method as an affordance for filtering dataframes
### Description

I found the following method extremally convenient in my code:
```python
def _locate(df:pl.DataFrame, **kwargs)->pl.DataFrame:
    '''Used to quickly find rows via equality checks.
    
    Example:
    # This
    df.locate(year=2020, quarter=2, metric='Sales')
    
    # is equivalent to 
    df.filter(pl.col('year')==2020).filter(pl.col('quarter')==2).filter(pl.col('metric')=='Sales')
    '''
            
    df_ans = df.filter(pl.all_horizontal([pl.col(k)==v for k,v in kwargs.items()]))
    
    return df_ans

pl.DataFrame.locate = _locate
```

I usualy would not recommend these methods be added to polars, but I was showing a friend the library on a clean installation and realized how much I would miss it if it were gone.

It allows you to write code like this:

```python
# Get the SignOn events for user #9
df_selected = df.locate(userId=9, eventType='SignOn')

# Equivalent to 
df_selected = df.filter(pl.col('userId')==9 & pl.col('eventType')=='SignOn')

```
Incidentally it also supports expressions:

```python
# Get the most recent SignOn date for user #9
df_selected = df.locate(userId=9, eventType='SignOn',eventDate=pl.col('eventDate').max().over('userId','eventType'))   
```

Thoughts? Is the task of retrieving a subset of rows via a few columns common enough to merit a shorthand?

### PR DESCRIPTION
**Pull request by mcrumiller**
Resolves #11642

This was mainly to get the ball rolling, I'm really looking forward to this feature. I'm happy to give this up to someone else that wants to do a more comprehensive implementation, but I think I covered what was discussed, and I did it using @alexander-beedie's suggested arguments. Suggestions are always welcome of course.

## Examples
From https://github.com/pola-rs/polars/issues/11642#issuecomment-1757183618:
```python
# just `*predicates`
df.filter( 
    pl.col("a") >= 0, 
    pl.col("b").is_not_null(),
    pl.col("").str.starts_with("!"),
 )

# supply list of predicates assembled elsewhere;
# becomes trivial to splat into filter
df.filter( *conditions )

# just `**constraints`
df.filter( 
    x = 0,
    y = "xxx",
    z = date.today(),
)

# mix of `*predicates` (via splat, for example), and `**constraints`
df.filter( *conditions, value=0, start_date=date.today() )
```

### CODE REVIEW

File: py-polars/polars/dataframe/frame.py [hunk 1]
```diff
@@ -3959,12 +3963,58 @@ def filter(
...
+        │ --- ┆ --- ┆ --- │
+        │ i64 ┆ i64 ┆ str │
+        ╞═════╪═════╪═════╡
+        │ 1   ┆ 6   ┆ a   │
+        └─────┴─────┴─────┘
+
         """
-        if _check_for_numpy(predicate) and isinstance(predicate, np.ndarray):
-            predicate = pl.Series(predicate)
+        has_predicates = len(predicates) > 0
```
Comment (mcrumiller) [inline py-polars/polars/dataframe/frame.py:3990]: FYI here, Ruff doesn't like `not not predicates`, but polars doesn't like `pl.Expr and x` for existence checks. Not sure if there's a slicker way to do this. Probably doesn't matter.

**Comment by gab23r** [comment]
Thanks for starting this, I think it is awesome ! I have two comments. 
I don't think that doing discrepancy between lazy and eager api is a good thing. I see many people that uses the eager api whereas lazy would be a better fit. If you protopyte things in eager it should be (as much as possible) trivial to convert it to lazy. 
Apart from that, being able give a list of values would be amazing i.e. `.filter(foo=[0,1])`

**Comment by alexander-beedie** [comment]
> I don't think this should be part of the lazy API (feel free to disagree)

I disagree - strongly 🤣 There is no reason to have this feature in one place but not the other; it doesn't impact laziness, so there is no obvious rationale the user could apply for understanding why it only works in one place but not the other.

The suggestion was to pattern `filter` after the same API (*args and/or **kwargs) that people are familiar with for `with_columns`, which works for both eager/lazy modes.

**Comment by mkleinbort-ic** [comment]
QQ: is the order of filters important? Or must constraints go after predicates?

```python
# Do both of these work?
df.filter(x=1, pl.col('y')>5) 
df.filter(pl.col('y')>5, x=1)
```

**Comment by cmdlineluser** [comment]
@mkleinbort-ic `df.filter(x=1, pl.col('y')>5) ` is a `SyntaxError` in Python.

Keyword args must be last.

**Comment by mcrumiller** [comment]
> I disagree - strongly, heh

Fair enough, I don't feel strongly about it. Will update.

**Comment by mcrumiller** [comment]
I also used a lot of mypy `#type: ignore` due to not feeling like trying to figure out the mypy `cast` conversions. If someone has a better handle on that, feel free to provide suggestions.

**Comment by alexander-beedie** [comment]
> If someone has a better handle on that, feel free to provide suggestions.

Il'l take a look later today ✌️

**Comment by mcrumiller** [comment]
Thanks Alex.

Does anyone else know why the Ubuntu doc tests seem to be incorrectly failing? Is this an issue we need to make?

```
File "/home/runner/work/polars/polars/py-polars/polars/dataframe/frame.py", line 39[30](https://github.com/pola-rs/polars/actions/runs/6496226583/job/17642723957#step:11:31), in polars.dataframe.frame.DataFrame.filter
Failed example:
    df.filter(pl.col("foo") < 3)
Expected:
    shape: (1, 3)    <--- THIS IS NOT WHAT WE SHOULD EXPECT!!
    ┌─────┬─────┬─────┐
    │ foo ┆ bar ┆ ham │
    │ --- ┆ --- ┆ --- │
    │ i64 ┆ i64 ┆ str │
    ╞═════╪═════╪═════╡
    │ 1   ┆ 6   ┆ a   │
    │ 2   ┆ 7   ┆ b   │
    └─────┴─────┴─────┘
Got:
    shape: (2, 3)
    ┌─────┬─────┬─────┐
    │ foo ┆ bar ┆ ham │
    │ --- ┆ --- ┆ --- │
    │ i64 ┆ i64 ┆ str │
    ╞═════╪═════╪═════╡
    │ 1   ┆ 6   ┆ a   │
    │ 2   ┆ 7   ┆ b   │
    └─────┴─────┴─────┘
```

**Comment by alexander-beedie** [comment]
> Does anyone else know why the Ubuntu doc tests seem to be incorrectly failing? Is this an issue we need to make?

Guess I'll take a look at that too 🤣

**Comment by alexander-beedie** [comment]
> Does anyone else know why the Ubuntu doc tests seem to be incorrectly failing? Is this an issue we need to make?

The tests were failing correctly; you were looking at the _lazy_ docstring, but the actual failure was coming from the _eager_ docstring where the "shape" really was wrong  ;)

**Comment by mcrumiller** [comment]
@alexander-beedie I don't quite understand the docstring failure (why does the provided expected shape not match the provided expected dataframe?) but I've also had a few whiskeys so that's me not understanding is probably what's most expected. But...thank you for the streamlined re-write, that looks much nicer 😄

**Comment by alexander-beedie** [comment]
> I don't quite understand the docstring failure (why does the provided expected shape not match the provided expected dataframe?) 

Screenshot 2023-10-13 at 10 10 59

It does now - that's the change I made above; I think you were just looking at the wrong docstring (because it was correct in lazy, but was _incorrect_ in eager, which `doctest` was pointing out)  ;)

File: py-polars/polars/dataframe/frame.py (hunk 1 above)
Comment (alexander-beedie) [reply py-polars/polars/dataframe/frame.py]: (Rewrote the logic a bit; avoids this issue)

**Comment by alexander-beedie** [comment]
> Apart from that, being able to give a list of values would be amazing i.e. `.filter(foo=[0,1])`

You'd want to represent that as `is_in`? 
Can't really do it, as then you couldn't filter for values in `List` dtype columns.

**Comment by gab23r** [comment]
Thanks for starting this, I think it is awesome ! I have two comments. 
I don't think that doing discrepancy between lazy and eager api is a good thing. I see many people that uses the eager api whereas lazy would be a better fit. If you protopyte things in eager it should be (as much as possible) trivial to convert it to lazy. 
Apart from that, being able give a list of values would be amazing i.e. `.filter(foo=[0,1])`

**Review by stinodego — CHANGES_REQUESTED** [review]
I'll review this more thoroughly later, but I can already say that this will need a special case to handle the input `predicate=...`. This should preserve existing behavior and throw a deprecation warning.

**Comment by alexander-beedie** [comment]
> I'll review this more thoroughly later, but I can already say that this will need a special case to handle the input `predicate=...`. This should preserve existing behavior and throw a deprecation warning.

[Indeed](https://github.com/pola-rs/polars/pull/11674#issuecomment-1760533516)  ;)

> ...need a migration strategy from predicate to *predicates on the off-chance somebody out there really has written df.filter(predicate=<xyz>), or they'll get it rewritten as a kwarg constraint...

Couldn't advance anything yesterday, but will update it shortly with deprecation handling.

**Comment by stinodego** [comment]
Right - I hadn't read the whole discussion 😅

**Comment by alexander-beedie** [comment]
(Superseded by #11740)

**Pull request closed without merging**